### DISCUSSION TITLE
Improve API regarding upper/lowercase arguments?

### DISCUSSION THREAD

**Discussion by peterkelm** [datenstrom/community#126] [Ask a question] [answered]
There seems to be a mix up of uppercase/lowercase in text string key naming convention in Yellow (0.6.1):

The language files (e.g. "language-en.ini") contain string like this: 

<pre>
LanguageDescription: English
</pre>

Note the **uppercase** writing in that language file. However, to successfully access this string a **lowercase** key name is needed:

<pre>
$yellow->text->getTextHtml("languageDescription")
</pre>

Is this expected behavior? I couldn't find that difference be documented anywhere.

(Just my 2ct - I'd prefer having these names be case sensitive.)

**Comment by wunderfeyd** [comment]
You are right.

There's a difference between the configuration/language keys and the array indices. The key names in the files are transformed during the loading process. I don't know the exact reason, but I think it was simply not seen.

From my point of view it's easier to modify all language and configuration files (yellow and yellow-extentions) than change every plugin accessing the arrays. What do you think? If you want to do it create a patch/pull request, otherwise such a change would be scheduled when refactoring the config and languages classes but it won't happen soon I'm afraid.

Your help is greatly appreciated.

**Comment by markseuffert** [comment]
Did you find anything else not working well in [Yellow API](https://datenstrom.se/yellow/help/api-for-developers)? 

Feedback from developers is greatly appreciated.

**Comment by peterkelm** [comment]
@wunderfeyd, I'd be okay with both approaches, as long as the behavior is logical and consistent. I'm pretty short on time currently but will try to look into it nevertheless.

@markseu: No, that was it for now. However, I haven't done much (more) with yellow over the past day or two.

**Comment by peterkelm** [comment]
As promised above I looked into this over the weekend. The functions:

<pre>function load($fileName)</pre>

in  "core.php" - one for loading data from the config file, the other to load strings from a language file - use "lcfirst" to convert the first character of all keys read into lowercase. The last occurrence is when page meta data is parsed.

<pre>function parseMeta($pageError = "")</pre>

In total "lcfirst" is used in five places inside "core.php" and it seems like removing them would be sufficient to get back to consistent array key names. Yet, there might have been a need for them that I am simply not aware of.

Does anyone know why the calls to "lcfirst" were introduced?

**Comment by markseuffert** [comment]
In the beginning there were many formats, for example there were different formats for the configuration file, content pages, PHP code, JSON code and query string. Now there are two main formats, which differentiate in upper/lower case. The internal meta data has always been starting with a lower case character, that's why there's a lcfirst when loading strings from file, to convert between formats.

While I agree with the problem, I don't have an opinion about the solution. There is not one format that will make everyone happy. If you are a non-developer you prefer this one, if you are a developer you prefer another one. I think this is the origin of the dilemma. From my point of view, we could change the internal meta data format to match the configuration file format, we could also use case-insensitive key names. 

Hope I could give some explanations. Any thoughts? Someone wants the Yellow API to be updated?

**Comment by markseuffert** [comment, marked as answer]
The Yellow API accepts different formats now, it was changed to work with case-insensitive key names. It doesn't matter anymore if meta data or text strings start with a lower case character or upper case character. For example this is possible now: `$yellow->text->getHtml("LanguageDescription")`.

Should be working now.
